src/succinctx/gateway.cairo

```diff
@@ -0,0 +1,277 @@
+    #[abi(embed_v0)]
+    impl OwnableImpl = ownable_cpt::OwnableImpl<ContractState>;
+    impl OwnableInternalImpl = ownable_cpt::InternalImpl<ContractState>;
+
+    #[storage]
+    struct Storage {
+        allowed_provers: LegacyMap<ContractAddress, bool>,
+        is_callback: bool,
+        nonce: u64,
+        requests: LegacyMap<u64, u256>,
+        verified_function_id: u256,
+        verified_input_hash: u256,
+        // TODO: change to Bytes after refactor succinctx
+        verified_output: u256,
+        #[substorage(v0)]
+        function_registry: function_registry_cpt::Storage,
+        #[substorage(v0)]
+        ownable: ownable_cpt::Storage,
+        #[substorage(v0)]
+        reentrancy_guard: ReentrancyGuardComponent::Storage
+    }
+
+    #[event]
+    #[derive(Drop, starknet::Event)]
+    enum Event {
+        RequestCall: RequestCall,
+        RequestCallback: RequestCallback,
+        #[flat]
+        FunctionRegistryEvent: function_registry_cpt::Event,
+        #[flat]
+        OwnableEvent: ownable_cpt::Event,
+        #[flat]
+        ReentrancyGuardEvent: ReentrancyGuardComponent::Event
+    }
+
+    #[derive(Drop, starknet::Event)]
+    struct RequestCall {
+        #[key]
+        function_id: u256,
+        input: Bytes,
+        entry_address: ContractAddress,
+        entry_calldata: Bytes,
+        entry_gas_limit: u32,
+        sender: ContractAddress,
+        fee_amount: u128,
+    }
+
+    #[derive(Drop, starknet::Event)]
+    struct RequestCallback {
+        #[key]
+        nonce: u64,
+        #[key]
+        function_id: u256,
+        input: Bytes,
+        context: Bytes,
+        callback_addr: ContractAddress,
+        callback_selector: felt252,
+        callback_gas_limit: u32,
+        fee_amount: u128,
+    }
+
+    mod Errors {
+        const INVALID_CALL: felt252 = 'Invalid call to verify';
+        const INVALID_REQUEST: felt252 = 'Invalid request for fullfilment';
+        const INVALID_PROOF: felt252 = 'Invalid proof provided';
+    }
+
+    #[constructor]
+    fn constructor(ref self: ContractState, owner: ContractAddress) {
+        self.ownable.initializer(owner);
+    }
+
+    #[abi(embed_v0)]
+    impl ISuccinctGatewayImpl of ISuccinctGateway<ContractState> {
+        /// Creates a onchain request for a proof. The output and proof is fulfilled asynchronously
+        /// by the provided callback.
+        ///
+        /// # Arguments
+        ///
+        /// * `function_id` - The function identifier.
+        /// * `input` - The function input.
+        /// * `context` - The function context.
+        /// * `callback_selector` - The selector of the callback function.
+        /// * `callback_gas_limit` - The gas limit for the callback function.
+        fn request_callback(
+            ref self: ContractState,
+            function_id: u256,
+            input: Bytes,
+            context: Bytes,
+            callback_selector: felt252,
```

Comment:
Please make this a u256 to make integration with Succinct easier

Reply:
As this is a callback selector for a cairo contract it was easier to keep as felt252.

If it's a pain point in integration we can look to convert.